codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2HeadersDecoder.java

```diff
@@ -152,4 +162,35 @@ protected final boolean validateHeaders() {
     protected Http2Headers newHeaders() {
         return new DefaultHttp2Headers(validateHeaders, (int) headerArraySizeAccumulator);
     }
+
+    protected void validate(Http2Headers headers) throws Http2Exception {
+        boolean isRequest = false;
+        boolean isResponse = false;
+
+        for (final Map.Entry<CharSequence, CharSequence> header : headers) {
+            final CharSequence key = header.getKey();
+            if (hasPseudoHeaderFormat(key)) {
+                final Http2Headers.PseudoHeaderName pseudoHeader = getPseudoHeader(key);
+                if (pseudoHeader == null) {
+                    throw connectionError(PROTOCOL_ERROR, "Invalid HTTP/2 pseudo-header '%s' encountered", key);
+                }
+
+                if (pseudoHeader.type() == Http2Headers.PseudoHeaderType.REQUEST) {
```

Comment:
This can be simplified:
```
Http2Headers.PseudoHeaderType prevType = null;
for (...) {
   ...
   if (pseudoHeader.type() != prevType)) {
      requestAndResponsePseudoHeadersMixed();
   }
   prevType = pseudoHeader.type();
}